Fix compiler errors

CromwellEnage opened the pull request:
<boost/graph/named_function_params.hpp>
* Use typename add_const<remove_reference<typename 'T'::value_type>::type>::type vice const typename T::value_type within boost::parameter::aux::tagged_argument type definition.
* Use boost::mpl::has_key vice boost::detail::parameter_exists.
* Add missing boost::graph::[function_name] overload definition that takes in ArgumentPack as the last parameter.

"test/Jamfile.v2"
* Mark msvc-14.0 failures as expected.

".travis.yml"
* Remove "sudo: false" because Travis Cl deprecated it.  See <https://blog.travis-ci.com/2018-10-04-combining-linux-infrastructures>.

anadon commented:
@jzmaddock This is over my pay grade.

jeremy-murphy commented on include/boost/graph/named_function_params.hpp:
```diff
@@ -364,7 +367,12 @@ BOOST_BGL_DECLARE_NAMED_PARAMS
+        , typename boost::add_const<
+              typename boost::remove_reference<typename T::value_type>::type
+          >::type
+      > tagged_arg_type;
```
Why do we add const and remove references from `value_type` now?

CromwellEnage replied:
I had compiler errors before without those type transformations.  I'm testing my local fork now to see if those errors still exist without them and, if so, to report what they are.

CromwellEnage replied:
It looks like I can get by without these type transformations.  I'm going to test my local fork with msvc 14.0 so I can document the regressions.

jeremy-murphy replied:
OK, keep us posted.

CromwellEnage replied:
These and other unnecessary changes have now been reverted.

jeremy-murphy commented on test/Jamfile.v2:
```diff
@@ -144,6 +129,35 @@ test-suite graph_test :
+        <toolset>msvc
+        <toolset-msvc:version>14.0
+        <address-model>64
+    ;
```
I'm not that familiar with this alias business; does this alias include only msvc, except version 14.0? Or does it include all toolchains except msvc 14.0?

CromwellEnage replied:
It includes all toolchains except msvc 14.0.

jeremy-murphy commented:
@jzmaddock John, I think we want to merge this asap.
There are a few details that I have questions about it but in general it looks good.

CromwellEnage commented:
On Sat, Dec 8, 2018 at 06:19 Jeremy W. Murphy <notifications@github.com>
wrote:

> *@jeremy-murphy* commented on this pull request.
> ------------------------------
>
> In test/Jamfile.v2
> <https://github.com/boostorg/graph/pull/137#discussion_r240001851>:
>
> > @@ -144,6 +129,35 @@ test-suite graph_test :
>      [ run delete_edge.cpp ]
>      ;
>
> +alias graph_test_no_msvc14 : :
> +        <toolset>msvc
> +        <toolset-msvc:version>14.0
> +        <address-model>64
> +    ;
>
> I'm not that familiar with this alias business; does this alias include
> only msvc, except version 14.0? Or does it include all toolchains except
> msvc 14.0?
>
The effect of the alias is to include all toolchains except msvc 14.0 for
those tests.


—
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/boostorg/graph/pull/137#pullrequestreview-182941232>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AfcJsP5eYgqmDgqHQHK8Hn6Sh_HCTB8Sks5u26BQgaJpZM4Ym1ID>
> .
>

jeremy-murphy commented:
> The effect of the alias is to include all toolchains except msvc 14.0 for those tests. —

Thanks for clarifying.

jzmaddock commented:
Doesn't this represent a regression for msvc-14?  For example astar_search_test.cpp develop builds with msvc-14.0, but not after this fix?

CromwellEnage commented:
Without the alias statement that excludes msvc 14.0, the astar_search_test and random_spanning_tree_test programs outputs exit status -1073741819, while the other affected tests encounter memory access violations at location 0xffffffff.  I'll look into these regressions later tonight.

jeremy-murphy commented:
John, yes, fair point, but some existing tests (isomorphism) don't even compile at the moment due to some change in the named parameter code.

OK, so maybe a better question is, when did the named parameter code break?

Thanks, Cromwell, I don't have a Windows platform to help with MSVC stuff.

jeremy-murphy commented:
@CromwellEnage can you give us the background on why/when this code is broken?

jeremy-murphy commented:
And, is there a possible alternative to just fix the isomorphism test and leave named parameters alone?

CromwellEnage commented:
On Sat, Dec 8, 2018 at 18:20 Jeremy W. Murphy <notifications@github.com>
wrote:

> @CromwellEnage <https://github.com/CromwellEnage> can you give us the
> background on why/when this code is broken?
>
> —
> You are receiving this because you were mentioned.
>
>
> Reply to this email directly, view it on GitHub
> <https://github.com/boostorg/graph/pull/137#issuecomment-445497350>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/AfcJsEJK8FmH5c9Q386OrkBJ9UYvRIwpks5u3EkugaJpZM4Ym1ID>
> .
>
In order to enable Boost.Parameter to support perfect forwarding, I had to
make breaking changes to boost::parameter::aux::tagged_argument and other
components that were not part of the public interface.  I made the effort
to ensure that all existing code that used *only* the public interface
still worked as before, so my changes were approved.

Unfortunately, this library and others relies on non-public portions of
Boost.Parameter, so now I’m focusing on submitting pull requests to either
switch to the public interface (preferred) or ensure correct usage of
current components (as is the case here).

CromwellEnage commented:
On Sat, Dec 8, 2018 at 18:36 Jeremy W. Murphy <notifications@github.com>
wrote:

> And, is there a possible alternative to just fix the isomorphism test and
> leave named parameters alone?
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/boostorg/graph/pull/137#issuecomment-445498168>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/AfcJsC9dmQrGs4JvVPbGGtT7PPE6YU0Vks5u3Ez9gaJpZM4Ym1ID>
> .
>
The isomorphism test wasn’t the only failure when I ran a fresh snapshot,
so alas I can’t leave named parameters alone.

jeremy-murphy commented:
Great, thanks for explaining the background.

CromwellEnage commented:
Zero progress on the regression-fixing front.  Not negative, not positive, zero.  I'm not receiving any clues as to why I'm going in the wrong direction, either.  It's frustrating, but I'll keep at it.

CromwellEnage commented:
I'll have to admit defeat on the msvc-14.0 regression front.

At any rate, the goal of this PR is to fix compiler errors that were occurring across the board.  I'd happily take up the regression issue again in another PR if I had more direct access to msvc-14.0 myself.

However, speaking of compiler errors, I saw some new ones popping up for Xcode, which I noted in the Jamfile.

jeremy-murphy commented:
On Wed., 12 Dec. 2018, 6:20 am Cromwell D. Enage <notifications@github.com
wrote:

> I'll have to admit defeat on the msvc-14.0 regression front.
>

Thanks for persisting this far. If I had MSVC I would help, but I'm using
Linux.

> At any rate, the goal of this PR is to fix compiler errors that were
> occurring across the board. I'd happily take up the regression issue again
> in another PR if I had more direct access to msvc-14.0 myself.
>
Yeah, let's stay focused on fixing the compilation errors and let the side
effects be what they are.
In my opinion it's better to have a single platform failing some tests than
a single test failing to even compile.
Presumably other libraries using named parameters might also be having
problems with msvc-14?
But wait, how have you been working on this if you don't have access to the
compiler yourself?

> However, speaking of compiler errors, I saw some new ones popping up for
> Xcode, which I noted in the Jamfile.
>
Huh, interesting. Seem to be some funky compilation issues coming to the
fore at the moment.

CromwellEnage commented:
On Tue, Dec 11, 2018 at 16:54 Jeremy W. Murphy <notifications@github.com>
wrote:

> On Wed., 12 Dec. 2018, 6:20 am Cromwell D. Enage <notifications@github.com
> wrote:
>
> > I'll have to admit defeat on the msvc-14.0 regression front.
> >
>
> Thanks for persisting this far. If I had MSVC I would help, but I'm using
> Linux.
>
> > At any rate, the goal of this PR is to fix compiler errors that were
> > occurring across the board. I'd happily take up the regression issue
> again
> > in another PR if I had more direct access to msvc-14.0 myself.
> >
> Yeah, let's stay focused on fixing the compilation errors and let the side
> effects be what they are.
> In my opinion it's better to have a single platform failing some tests than
> a single test failing to even compile.
> Presumably other libraries using named parameters might also be having
> problems with msvc-14?
> But wait, how have you been working on this if you don't have access to the
> compiler yourself?


I’ve been doing the main work on the develop branch of my own Github fork
and testing that fork using AppVeyor.  Each run is rather time-consuming.
When I’m ready to publish, I copy a snapshot to the PR branch.


> > However, speaking of compiler errors, I saw some new ones popping up for
> > Xcode, which I noted in the Jamfile.
> >
> Huh, interesting. Seem to be some funky compilation issues coming to the
> fore at the moment.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/boostorg/graph/pull/137#issuecomment-446376756>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/AfcJsE3mcmWORAXWah-NHUcQedHYlULVks5u4CmEgaJpZM4Ym1ID>
> .
>

jeremy-murphy commented:
Btw, I actually had to look up "vice" to realize what it means (and how it is pronounced) in that context.  :)  I don't think that particular usage is common in Australia.

CromwellEnage commented:
On Wed, Dec 12, 2018 at 07:21 Jeremy W. Murphy <notifications@github.com>
wrote:

> Btw, I actually had to look up "vice" to realize what it means (and how it
> is pronounced) in that context. :) I don't think that particular usage is
> common in Australia.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/boostorg/graph/pull/137#issuecomment-446569033>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/AfcJsDeNRZG_oPnlcw0zD5TERby2RWwXks5u4PTZgaJpZM4Ym1ID>
> .
>
My usage of “vice” wasn’t common for me either until I served in the U. S.
Navy.

jzmaddock commented:
Is the code now stable and ready to go?

If so I can look at the msvc failure locally.

CromwellEnage commented:
On Wed, Dec 12, 2018 at 13:20 jzmaddock <notifications@github.com> wrote:

> Is the code now stable and ready to go?
>
> If so I can look at the msvc failure locally.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/boostorg/graph/pull/137#issuecomment-446690181>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/AfcJsIUiKbXusK8cf-IcOvTCkhnUiJB0ks5u4UkHgaJpZM4Ym1ID>
> .
>
The code is good to go on my end.

jzmaddock commented:
Well... I don't understand the code it seems :(

Here's what I have found:

In boyer_myrvold_planar_test.hpp: 78 the access violation occurs on:

```
        boyer_myrvold_impl
          <graph_t, 
           vertex_index_map_t,
           graph::detail::no_old_handles,
           graph::detail::no_embedding
          >
          planarity_tester(args[graph], 
                           args[vertex_index_map | 
                                get(vertex_index, args[graph])
                                ]
                           );
```
The subscript calls to args[] both succeed, but then the violation occurs loading the parameters to planarity_tester on the stack.

Changing the above to:
```
        auto a1 = args[graph];
        auto a2 = args[vertex_index_map |
           get(vertex_index, args[graph])
        ];

        boyer_myrvold_impl
          <graph_t, 
           vertex_index_map_t,
           graph::detail::no_old_handles,
           graph::detail::no_embedding
          >
          planarity_tester(a1,
                           a2
                           );

        return planarity_tester.is_planar() ? true : false;
      }
```

Fixes the first few calls, but then we get a new access violation on

```
auto a2 = args[vertex_index_map |
           get(vertex_index, args[graph])
        ];
```
The "obvious" thing to try then is to replace `args[graph]` with `a1`, but this leads to an impenetrable compiler error I don't understand.
In any case it looks as if one or more results from these calls is a dangling reference - can that be possible?  What happens if you pass the tests through clang's asan/ubsan?

CromwellEnage commented:
Progress!

I've replaced the first 'auto' with a const reference to the graph type, which can be extracted from the argument pack via parameter::value_type.  I've replaced the second 'auto' with the vertex index map type of the graph, *not* reference qualified.  Same deal for the edge index map.

If the get() call returns by value, then the bracket operator of args will treat it as an rvalue reference.  Therefore, msvc-14.0 may be evaluating the second 'auto' such that the object holds a dangling reference.

jeremy-murphy commented:
That's great.

jzmaddock commented:
Closer, but no cigar just yet.
That now fails on:

```
            vertex_index_map_t i_map = args[vertex_index_map | get(vertex_index, g)];
```

But only when the get() call returns something with internal state (ie not just a "default" tag).
It's fixed by using:
```
            graph_t const& g = args[graph];
            auto a = get(vertex_index, g);
            vertex_index_map_t i_map = args[vertex_index_map | a];
```
The difference is that operator | no longer calls the rvalue-ref overload, but I don't see why that's an issue at present :(
Hopefully you can see what actual type should replace the auto above?

jzmaddock commented:
>All right, now I've applied our methods to astar_search and random_spanning_tree. They work in my fork.

Great!  Just waiting for CI to cycle, there's quite a bit backed up, so it may take a while :(

CromwellEnage commented:
I'm happy on my end.

jzmaddock commented:
OK, merging.

jeremy-murphy commented:
Great stuff, thanks for the coordinated effort and particularly to Cromwell for the obscure code changes that would have taken the rest of us I don't know how long to figure out.

CromwellEnage commented:
When a boost::parameter::keyword object such as _isomorphism_map is assigned an rvalue reference such as the object returned by make_assoc_property_map(mapping), it returns a tagged_argument_rref rather than a tagged_argument.  I'll open up another PR that adds clang++-6 to the test matrix and fixes this issue.

jzmaddock commented:
>I'll open up another PR that adds clang++-6 to the test matrix and fixes this issue.

I only happened to give clang as an example - the test matrix has that test failing with nearly all compilers, including msvc and gcc.  As I said, I'm lost as to why CI doesn't show the failure.

CromwellEnage commented:
It looks like my fix only worked for Cl because the compilers there were able to find the new overload that I added to named_function_params.hpp, while the compilers used in the Boost develop test matrix did not...which I also find strange.  Regardless, I'm going to work on a more robust fix.
